lib/AST/TypeJoinMeet.cpp

```diff
@@ -32,6 +32,18 @@ Type Type::join(Type type1, Type type2) {
   if (type1->isEqual(type2))
     return type1;
 
+  // If both are metatypes, compute the join of the instance type and wrap the
+  // result in a metatype.
+  if (auto *metatype1 = type1->getAs<MetatypeType>()) {
+    if (auto *metatype2 = type2->getAs<MetatypeType>()) {
+      auto result = Type::join(metatype1->getInstanceType(),
+                               metatype2->getInstanceType());
+      if (!result)
+        return result;
+      return MetatypeType::get(result);
```

Comment:
It's not always true that if `T < U`, then`T.Type < U.Type` in our type system. Let's whitelist the cases where it's true:

* Both instance types `mayHaveSuperclass()`, or
* Both are existential metatypes

and not compute a join in the other cases.